### PR TITLE
feat(bulk-load): group bulk load request in replica group

### PR DESCRIPTION
**Pull request by hycdong**
This pull request is about primary broadcast group_bulk_load_request to secondaries.
- primary replica receive bulk_load_request (`on_bulk_load` #457)
- primary replica broadcast group_bulk_load_request (`broadcast_group_bulk_load`)
- secondary replica calls `on_group_bulk_load` when receive group_bulk_load_request
  - validate replica status and ballot
  - execute actual bulk load process (do_bulk_load - not implemented in this pr )
  - report bulk load states to primary (report_bulk_load_states_to_primary - not implemented in this pr)
- primary replica calls `on_group_bulk_load_reply` when receive group_bulk_load_response, part-implemented in this pr

Whole bulk load process document can be found in official website: [bulk_load_design](https://pegasus-kv.github.io/2020/02/18/bulk-load-design.html)

### CODE REVIEW

File: src/dist/replication/lib/replica_bulk_load.cpp [hunk 1]
```diff
@@ -71,7 +73,115 @@ void replica::on_bulk_load(const bulk_load_request &request, /*out*/ bulk_load_r
 // ThreadPool: THREAD_POOL_REPLICATION
 void replica::broadcast_group_bulk_load(const bulk_load_request &meta_req)
 {
+    if (!_primary_states.learners.empty()) {
+        dwarn_replica("has learners, skip broadcast group bulk load request");
+        return;
+    }
+
+    if (_primary_states.group_bulk_load_pending_replies.size() > 0) {
```
Comment (acelyc111) [inline src/dist/replication/lib/replica_bulk_load.cpp:81]:
```suggestion
    if (!_primary_states.group_bulk_load_pending_replies.empty()) {
```

File: src/dist/replication/lib/replica_bulk_load.cpp [hunk 2]
```diff
@@ -71,7 +73,115 @@ void replica::on_bulk_load(const bulk_load_request &request, /*out*/ bulk_load_r
...
+                                 /*out*/ group_bulk_load_response &response)
+{
+    _checker.only_one_thread_access();
+
+    response.err = ERR_OK;
+
+    if (request.config.ballot < get_ballot()) {
+        response.err = ERR_VERSION_OUTDATED;
+        dwarn_replica(
+            "receive outdated group_bulk_load request, request ballot({}) VS loca ballot({})",
```
Comment (acelyc111) [inline src/dist/replication/lib/replica_bulk_load.cpp:126]:
```suggestion
            "receive outdated group_bulk_load request, request ballot({}) VS local ballot({})",
```

File: src/dist/replication/lib/replica_bulk_load.cpp [hunk 3]
```diff
@@ -71,7 +73,115 @@ void replica::on_bulk_load(const bulk_load_request &request, /*out*/ bulk_load_r
...
+            request.config.ballot,
+            get_ballot());
+        return;
+    }
+    if (request.config.ballot > get_ballot()) {
+        response.err = ERR_INVALID_STATE;
+        dwarn_replica("receive group_bulk_load request, local ballot is outdated, request "
+                      "ballot({}) VS loca ballot({})",
```
Comment (acelyc111) [inline src/dist/replication/lib/replica_bulk_load.cpp:134]:
```suggestion
                      "ballot({}) VS local ballot({})",
```

File: src/dist/replication/lib/replica_context.cpp [hunk 4]
```diff
@@ -94,6 +94,14 @@ void primary_context::cleanup(bool clean_pending_mutations)
     // clean up register child task
     CLEANUP_TASK_ALWAYS(register_child_task)
 
+    // cleanup group bulk load
+    for (auto it = group_bulk_load_pending_replies.begin();
```
Comment (acelyc111) [inline src/dist/replication/lib/replica_context.cpp:98]: Preffer to use range loop

File: src/dist/replication/lib/replica_bulk_load.cpp [hunk 5]
```diff
@@ -71,7 +73,115 @@ void replica::on_bulk_load(const bulk_load_request &request, /*out*/ bulk_load_r
...
+        return;
+    }
+
+    if (_primary_states.group_bulk_load_pending_replies.size() > 0) {
+        dwarn_replica("{} group bulk_load replies are still pending, cancel it firstly",
+                      static_cast<int>(_primary_states.group_bulk_load_pending_replies.size()));
+        for (const auto &kv : _primary_states.group_bulk_load_pending_replies) {
+            kv.second->cancel(true);
+        }
+        _primary_states.group_bulk_load_pending_replies.clear();
```
Comment (acelyc111) [inline src/dist/replication/lib/replica_bulk_load.cpp:87]: Does it mean we will resend group_bulk_load_request to sendaries? How does secondary replica deal with duplicate bulk load request?

Comment (hycdong) [reply src/dist/replication/lib/replica_bulk_load.cpp:87]: - meta -> primary: send bulk _load_request
- primary -> secondary: send group_bulk_load_request (request 1)
- secondary -> primary : report bulk load state
- primary -> meta: report group bulk load states
- if bulk load not finished, meta send bulk_load_request again
- primary -> secondary: send group_bulk_load_request (request 2)

Primary will not resend group_bulk_load_request, request 1 and request 2 are different.
When primary tries to send request 2, primary will check if there are request 1 pending firstly, if there are pending request 1, primary will cancel tasks and send request 2.
Meta server will not send bulk_load_request too soon, the interval is longer than 5 seconds (will be shown in further pull request), it is safe to cancel the long-time, pending request.

File: src/dist/replication/lib/replica_bulk_load.cpp [hunk 6]
```diff
@@ -71,7 +73,115 @@ void replica::on_bulk_load(const bulk_load_request &request, /*out*/ bulk_load_r
...
+    if (!_primary_states.learners.empty()) {
+        dwarn_replica("has learners, skip broadcast group bulk load request");
+        return;
+    }
+
+    if (!_primary_states.group_bulk_load_pending_replies.empty()) {
+        dwarn_replica("{} group bulk_load replies are still pending, cancel it firstly",
+                      static_cast<int>(_primary_states.group_bulk_load_pending_replies.size()));
+        for (const auto &kv : _primary_states.group_bulk_load_pending_replies) {
+            kv.second->cancel(true);
```
Comment (levy5307) [inline src/dist/replication/lib/replica_bulk_load.cpp:85]: CLEANUP_TASK_ALWAYS(kv.second);

File: src/dist/replication/lib/replica_bulk_load.cpp [hunk 7]
```diff
@@ -71,7 +73,115 @@ void replica::on_bulk_load(const bulk_load_request &request, /*out*/ bulk_load_r
...
 void replica::broadcast_group_bulk_load(const bulk_load_request &meta_req)
 {
+    if (!_primary_states.learners.empty()) {
+        dwarn_replica("has learners, skip broadcast group bulk load request");
+        return;
+    }
+
+    if (!_primary_states.group_bulk_load_pending_replies.empty()) {
+        dwarn_replica("{} group bulk_load replies are still pending, cancel it firstly",
+                      static_cast<int>(_primary_states.group_bulk_load_pending_replies.size()));
```
Comment (neverchanje) [inline src/dist/replication/lib/replica_bulk_load.cpp:83]: no need to static_cast
```suggestion
                      _primary_states.group_bulk_load_pending_replies.size());
```

File: src/dist/replication/lib/replica_bulk_load.cpp (hunk 6 above)
Comment (neverchanje) [reply src/dist/replication/lib/replica_bulk_load.cpp:85]: `cancel(false)`

`cancel(true)` will wait until the rpc task finishes, and blocks the REPLICATION thread.

**Pull request merged**